src/subscription/subscribe.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2017, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import { isAsyncIterable } from 'iterall';
+import {
+  addPath,
+  buildExecutionContext,
+  collectFields,
+  defaultFieldResolver,
+  execute,
+  getFieldDef,
+  getOperationRootType,
+  buildResolveInfo,
+  resolveFieldValueOrError,
+} from '../execution/execute';
+import { GraphQLSchema } from '../type/schema';
+import invariant from '../jsutils/invariant';
+import mapAsyncIterator from './mapAsyncIterator';
+
+import type {
+  ExecutionContext,
+  ExecutionResult,
+} from '../execution/execute';
+import type {
+  DocumentNode,
+  OperationDefinitionNode,
+} from '../language/ast';
+
+
+/**
+ * Implements the "Subscribing to request" section of the GraphQL specification.
+ *
+ * Returns an AsyncIterator
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function subscribe(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  rootValue?: mixed,
+  contextValue?: mixed,
+  variableValues?: ?{[key: string]: mixed},
+  operationName?: ?string,
+): AsyncIterator<ExecutionResult> {
+  // If a valid context cannot be created due to incorrect arguments,
+  // this will throw an error.
+  const exeContext = buildExecutionContext(
+    schema,
+    document,
+    rootValue,
+    contextValue,
+    variableValues,
+    operationName
+  );
+
+  // Call the `subscribe()` resolver or the default resolver to produce an
+  // AsyncIterable yielding raw payloads.
+  const subscription = resolveSubscription(
+    exeContext,
+    exeContext.operation,
+    rootValue
+  );
+
+  // For each payload yielded from a subscription, map it over the normal
+  // GraphQL `execute` function, with `payload` as the rootValue.
+  return mapAsyncIterator(
+    subscription,
+    payload => execute(
+      schema,
+      document,
+      payload,
+      contextValue,
```

Comment:
Can there be some way to customize this context, possibly by adding some way to preprocess the arguments to `execute` here? (This could go well with converting these function signatures to take objects, per https://github.com/graphql/graphql-js/issues/356.)

In our current implementation, we use time-limited auth tokens. For our subscription implementation, we allow the client to update the auth token used.

This auth token is bound to the client session (it comes from somewhere separate). Implementation-wise, we currently swap out the context used for resolving subscriptions, which allows us to continue to use an up-to-date auth token without e.g. tearing down and recreating all the subscriptions.